cgen: more complex expression in ternary operator

OS: macos, macOS, 12.5.1, 21G83
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i7-8569U CPU @ 2.80GHz
CC version: Apple clang version 13.1.6 (clang-1316.0.21.2.5)

getwd: /Users/shove
vmodules: /Users/shove/.vmodules
vroot: /usr/local/v
vexe: /usr/local/v/v
vexe mtime: 2022-09-05 00:52:44
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.1 d0d5f1d

Git version: git version 2.32.1 (Apple Git-133)
Git vroot status: 0.3-460-gd0d5f1d4
.git/config present: true
thirdparty/tcc status: thirdparty-macos-amd64 46662e20

**What did you do?**

```v
import rand

fn main() {
	a := [1,2,3]

	b := if true { a[rand.intn(a.len) or { 0 }] } else { 0 }
	println(b)
}
```

**What did you expect to see?**

There is no bug, or the checker report does not allow this syntax.
 
**What did you see instead?**

```
==================
   ^~~~~~~~~~~~~~~
/tmp/v_501/a.12421870952442269733.tmp.c:15283:33: error: expected ')'
        int b = (true ? (               _option_int _t1 = rand__intn(a.len);
                                                    ^
/tmp/v_501/a.12421870952442269733.tmp.c:15283:18: note: to match this '('
        int b = (true ? (               _option_int _t1 = rand__intn(a.len);
                        ^
/tmp/v_501/a.12421870952442269733.tmp.c:15283:56: error: expected expression
        int b = (true ? (               _option_int _t1 = rand__intn(a.len);
                                                                           ^
/tmp/v_501/a.12421870952442269733.tmp.c:15283:56: error: expected ':'
        int b = (true ? (               _option_int _t1 = rand__intn(a.len);
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

C code:

```c
VV_LOCAL_SYMBOL void main__main(void) {
	Array_int a = new_array_from_c_array_noscan(3, 3, sizeof(int), _MOV((int[3]){1, 2, 3}));
	int b = (true ? (		_option_int _t1 = rand__intn(a.len);
		if (_t1.state != 0) {
			IError err = _t1.err;
			*(int*) _t1.data = 0;
		}
		
 (*(int*)array_get(a,  (*(int*)_t1.data)))) : (0));
	println(int_str(b));
}

```

It should be like this:

```c
VV_LOCAL_SYMBOL void main__main(void) {
	Array_int a = new_array_from_c_array_noscan(3, 3, sizeof(int), _MOV((int[3]){1, 2, 3}));
	_option_int _t1 = rand__intn(a.len);
		if (_t1.state != 0) {
			IError err = _t1.err;
			*(int*) _t1.data = 0;
		}
	int b = (true ? ((*(int*)array_get(a,  (*(int*)_t1.data)))) : (0));
	println(int_str(b));
}

```

--- a/vlib/v/gen/c/if.v
+++ b/vlib/v/gen/c/if.v
@@ -60,6 +60,14 @@ fn (mut g Gen) need_tmp_var_in_expr(expr ast.Expr) bool {
 			}
 		}
 	}
+	if expr is ast.IndexExpr {
+		if expr.or_expr.kind != .absent {
+			return true
+		}
+		if g.need_tmp_var_in_expr(expr.index) {
+			return true
+		}
+	}
 	return false
 }
 

--- a/vlib/v/tests/if_expr_with_index_expr_test.v
+++ b/vlib/v/tests/if_expr_with_index_expr_test.v
@@ -1,0 +1,9 @@
+import rand
+
+fn test_if_expr_with_index_expr() {
+	a := [1, 2, 3]
+
+	b := if true { a[rand.intn(a.len) or { 0 }] } else { 0 }
+	println(b)
+	assert true
+}

